### PR TITLE
feat(vlm): add Gemma 4 multimodal support

### PR DESCRIPTION
**Pull request by Yanko-7**
## Motivation

This PR redirects the final implementation stage after #1481 onto the current stack. It adds **Gemma 4** support and carries the runtime/scheduler follow-ups needed by the expanded multimodal model set.

| Order | PR | Scope | Status |
| --- | --- | --- | --- |
| 1 | #1479 | Cache-aware multimodal scheduling | Prerequisite |
| 2 | #1480 | Shared runtime and Qwen2.5-VL | Required |
| 3 | #1481 | Qwen3-VL, MiMo-V2, and Qwen3.5 | Required |
| 4 | **#1482** | Gemma 4 and runtime/scheduler follow-ups | This PR |

The stacked PRs continue to target `epic/vlm`, so GitHub displays the cumulative stack. The logical diff for this PR is the seven implementation commits after #1481.

### Gemma 4 model and configuration

- Add local Gemma 4 text and vision configuration support.
- Register `Gemma4ForConditionalGeneration` as an in-model multimodal architecture.
- Implement the Gemma 4 decoder, hybrid global/sliding-window attention behavior, RoPE variants, logits path, and checkpoint mappings.
- Preserve per-layer attention head dimensions and cache dtype behavior.

### Native Gemma 4 vision tower

- Add patch projection, learned position embeddings, interleaved 2-D rotary embeddings, Q/K normalization, encoder blocks, pooling, and multimodal projection.
- Use the shared varlen vision attention implementation and DP/TP vision sharding.
- Support bucketed precompile and packed output-capacity reporting.
- Restore encoded rows to the item-ordered array consumed by the shared merge and embedding pool.

### Gemma 4 processor

- Add `Gemma4Processor` for pixel features, image position metadata, placeholder lengths, and multimodal item construction.
- Validate processor output against the request's image batch and pooling configuration.

### Multimodal preprocessing and model integration

- Move processor work off the asyncio event loop through a bounded executor.
- Share the asynchronous preprocessing path across Qwen-VL, MiMo-V2, and Gemma 4.
- Enable the compatible fast Qwen image processor.
- Consolidate model-side mRoPE vision execution on `run_mrope_vision_model`.
- Route scheduler-side multimodal processing through the new processor interface.

### Scheduler and cache safety

- Reserve the complete remaining output budget for running ignore-EOS requests.
- Balance queued work across DP ranks.
- Apply page-aligned safety checks to hybrid allocation.
- Reserve sufficient SWA cache headroom to avoid avoidable decode retraction.

## Testing

```bash
python -m pytest -q \
  python/sgl_jax/test/models/test_gemma4_vision.py \
  python/sgl_jax/test/multimodal/test_gemma4_processor.py \
  python/sgl_jax/test/multimodal/test_qwen_vl_processor.py \
  python/sgl_jax/test/test_model_runner_kv_cache_mixin.py \
  python/sgl_jax/test/test_swa_schedule_budget.py
```

## Review Notes

- Review the seven implementation commits after #1481.
- Adapter work for Qwen3-VL, MiMo-V2, and Qwen3.5 belongs to #1481.
- Shared runtime and Qwen2.5-VL behavior belongs to #1480.
- Qwen3-Omni remains outside the scope of this stack.

### CODE REVIEW

**Comment by Yanko-7** [comment]
Implemented by https://github.com/sgl-project/sglang-jax/pull/1659
